src/percy-agent-client/dom.ts

```diff
@@ -0,0 +1,205 @@
+
+    return (
+      `<!DOCTYPE ${doctype.name}` + publicDeclaration + systemDeclaration + '>'
+    )
+  }
+
+  /**
+   * Takes the raw DOM from the snapshot, performs necessary mutations,
+   * and finally returns a deep clone of the original DOM.
+   *
+   */
+  private cloneDOM(): HTMLDocument {
+    this.mutateOriginalDOM()
+    let clone = this.originalDOM.cloneNode(true)
+    clone = this.stabilizeDOM(clone)
+
+    return clone
+  }
+
+  /**
+   * Serialize parts of the DOM that aren't preserved in a deep clone. Anything
+   * that is not encoded in the DOM tree needs to be accounted for here (CSSOM,
+   * input values, canvas, etc).
+   *
+   * This method should always capture these values from the origial DOM and
+   * serialize into the cloned DOM. Never mutate the origial DOM.
+   *
+   */
+  private stabilizeDOM(clonedDOM: any): HTMLElement {
+    let stabilizedDOMClone
+    stabilizedDOMClone = this.serializeInputElements(clonedDOM)
+
+    // We only want to serialize the CSSOM if JS isn't enabled.
+    if (!this.options.enableJavaScript) {
+      stabilizedDOMClone = this.serializeCssOm(clonedDOM)
+    }
+
+    return stabilizedDOMClone.documentElement
+  }
+
+  /**
+   * Capture in-memory form element values and serialize those values to their
+   * respective elements in the cloned DOM.
+   *
+   * Form element values for non-controlled elements won't be captured with a
+   * domClone. To fix this, we explicitly set the `value` attribute on those
+   * form controls to make sure they persist in snapshots.
+   *
+   */
+  private serializeInputElements(clonedDOM: HTMLDocument): HTMLDocument {
+    const formNodes = this.originalDOM.querySelectorAll('input, textarea')
+    const formElements = Array.from(formNodes)
+
+    formElements.forEach((elem: any) => {
+      const inputId = elem.attributes['data-percy-element-id'].value
+      const selector = `[data-percy-element-id="${inputId}"]`
+      const cloneEl = clonedDOM.querySelector(selector) as HTMLInputElement
+
+      switch (elem.type) {
+        case 'checkbox':
+        case 'radio':
+          if (elem.checked && !elem.hasAttribute('checked')) {
+            cloneEl!.setAttribute('checked', '')
+          }
+          break
+        case 'textarea':
+          // setting text or value does not work but innerText does
+          if (elem.innerText !== elem.value) {
+            cloneEl!.innerText = elem.value
+          }
+        default:
+          if (!elem.getAttribute('value')) {
+            cloneEl!.setAttribute('value', elem.value)
+          }
+      }
+    })
+
+    return clonedDOM
+  }
+
+  /**
+   * Capture in-memory styles & serialize those styles into the cloned DOM.
+   *
+   * Without this, applications that use packages like `styled-components` will be
+   * missing styles and appear broken. The CSSOM provides a way to create CSS
+   * that only lives in memory (not an asset or in the DOM).
+   *
+   */
+  private serializeCssOm(documentClone: any) {
```

Comment:
It'd be nice to do some perf testing here (re #188). I _think_ we're going to have to bite the perf cost here of iterating through each in-memory stylesheet each time. 

Maybe in the future we can do some caching. Feels dangerous though, I'd rather capture the CSSOM each time to be 100% sure it's correct.